src/init.cpp

```diff
@@ -1089,13 +1091,23 @@ static bool LockDataDirectory(bool probeOnly)
     return true;
 }
 
-bool AppInitSanityChecks()
+bool AppInitSanityChecks(const kernel::Context& kernel)
 {
     // ********************************************************* Step 4: sanity checks
-
-    init::SetGlobals();
-
-    if (!init::SanityChecks()) {
+    auto maybe_error = kernel::SanityChecks(kernel);
+
+    if (maybe_error.has_value()) {
+        switch (maybe_error.value()) {
+        case kernel::SanityCheckError::ERROR_ECC:
+            InitError(Untranslated("Elliptic curve cryptography sanity check failure. Aborting."));
+            break;
+        case kernel::SanityCheckError::ERROR_RANDOM:
+            InitError(Untranslated("OS cryptographic RNG sanity check failure. Aborting."));
+            break;
+        case kernel::SanityCheckError::ERROR_CHRONO:
+            InitError(Untranslated("Clock epoch mismatch. Aborting."));
+            break;
```

Comment:
Addressed in d87784ac87364fc977bbf9769c8bdb72dea8cbf9, feel free to reopen this convo if not resolved!

Reply:
FWIW, gcc/clang have built-ins (`__builtin_unreachable`) for [expressing this explicitly](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html).  

LLVM even has a [macro for it](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/ErrorHandling.h#L143), and specifies the use of `llvm_unreachable` in its [coding standard](https://llvm.org/docs/CodingStandards.html).

Maybe we should borrow the idea and add a similar macro. Not in this PR though, of course :)

Reply:
Interesting, though I think we'd always want to abort, even in release builds. For example, it would be UB to skip `return` in a non-void function, no? This is also what the docs say:

> llvm_unreachable() becomes an optimizer hint that the current location is not supposed to be reachable: the hint turns such code path into undefined behavior.

https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/ErrorHandling.h#L131

For reference, in https://github.com/bitcoin/bitcoin/pull/24812#discussion_r851597471 I argued that `assert(0)` is good enough for us. Also, I don't think there is a compiler that doesn't understand `assert(0)`?

Reply:
I think the primary question here was why there is no default case, to which the answer is that we happily accept a fallthrough to an InitError. No objection to a macro that just calls `assert(0)`, if ppl think it is useful.

Reply:
> ... why there is no default case ...

My favorite reason is that, because the switch value is an enum, we get a compile-time warning (or error with `-Werror` which is used by CI) if not all enum values are listed.